lib/statemachine/group.go

```diff
@@ -0,0 +1,116 @@
+package statemachine
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/filecoin-project/go-statestore"
+	"github.com/ipfs/go-datastore"
+	"golang.org/x/xerrors"
+)
+
+type StateHandler interface {
+	// returns
+	Plan(events []Event, user interface{}) (interface{}, error)
+}
+
+// StateGroup manages a group of state machines sharing the same logic
+type StateGroup struct {
+	sts       *statestore.StateStore
+	hnd       StateHandler
+	stateType reflect.Type
+
+	lk  sync.Mutex
+	sms map[datastore.Key]*StateMachine
+}
+
+// stateType: T - (reflect.TypeOf(MyStateStruct{}))
+func New(ds datastore.Datastore, hnd StateHandler, stateType reflect.Type) *StateGroup {
```

Comment:
Can we just pass the empty struct here?
This way you don't have to use reflect everywhere else.